File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/QifImportDialogFragment.java
package org.totschnig.myexpenses.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.database.MergeCursor;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.SimpleCursorAdapter;
import android.widget.Spinner;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.QifImport;
import org.totschnig.myexpenses.adapter.CurrencyAdapter;
import org.totschnig.myexpenses.export.qif.QifDateFormat;
import org.totschnig.myexpenses.model.ExportFormat;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.util.UiUtils;
import org.totschnig.myexpenses.util.Utils;
import org.totschnig.myexpenses.viewmodel.CurrencyViewModel;
import org.totschnig.myexpenses.viewmodel.data.Currency;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.lifecycle.ViewModelProvider;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;

import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ACCOUNTID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CURRENCY;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_LABEL;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ROWID;
import static org.totschnig.myexpenses.provider.DatabaseConstants.KEY_SEALED;

public class QifImportDialogFragment extends TextSourceDialogFragment implements
    LoaderManager.LoaderCallbacks<Cursor>, OnItemSelectedListener {
  Spinner mAccountSpinner, mDateFormatSpinner, mCurrencySpinner, mEncodingSpinner;
  private SimpleCursorAdapter mAccountsAdapter;

  public static final String PREF_KEY_IMPORT_DATE_FORMAT = "import_qif_date_format";
  public static final String PREF_KEY_IMPORT_ENCODING = "import_qif_encoding";
  private MergeCursor mAccountsCursor;
  private long accountId = 0;
  private String currency = null;
  private CurrencyViewModel currencyViewModel;

  public static QifImportDialogFragment newInstance() {
    return new QifImportDialogFragment();
  }

  @Override
  public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    currencyViewModel = new ViewModelProvider(this).get(CurrencyViewModel.class);
  }

  @NonNull
  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    if (savedInstanceState != null) {
      accountId = savedInstanceState.getLong(KEY_ACCOUNTID);
      currency = savedInstanceState.getString(KEY_CURRENCY);
    }
    return super.onCreateDialog(savedInstanceState);
  }

  @Override
  protected int getLayoutId() {
    return R.layout.import_dialog;
  }

  @Override
  protected String getLayoutTitle() {
    return getString(R.string.pref_import_title,
        getFormat().name());
  }

  private ExportFormat getFormat() {
    return ExportFormat.QIF;
  }

  @Override
  public String getTypeName() {
    return getFormat().name();
  }

  @Override
  public String getPrefKey() {
    return "import_" + getFormat().getExtension() + "_file_uri";
  }

  @Override
  public void onClick(DialogInterface dialog, int id) {
    if (getActivity() == null) {
      return;
    }
    if (id == AlertDialog.BUTTON_POSITIVE) {
      QifDateFormat format = (QifDateFormat) mDateFormatSpinner.getSelectedItem();
      String encoding = (String) mEncodingSpinner.getSelectedItem();
      maybePersistUri();
      prefHandler.putString(PREF_KEY_IMPORT_ENCODING, encoding);
      prefHandler.putString(PREF_KEY_IMPORT_DATE_FORMAT, format.name());
      ((QifImport) getActivity()).onSourceSelected(
          mUri,
          format,
          mAccountSpinner.getSelectedItemId(),
          ((Currency) mCurrencySpinner.getSelectedItem()).getCode(),
          mImportTransactions.isChecked(),
          mImportCategories.isChecked(),
          mImportParties.isChecked(),
          encoding
      );
    } else {
      super.onClick(dialog, id);
    }
  }

  @NonNull
  @Override
  public Loader<Cursor> onCreateLoader(int id, Bundle args) {
    if (getActivity() == null) {
      return null;
    }
    return new CursorLoader(
        getActivity(),
        TransactionProvider.ACCOUNTS_BASE_URI,
        new String[]{
            KEY_ROWID,
            KEY_LABEL,
            KEY_CURRENCY},
        KEY_SEALED + " = 0 ", null, null);
  }

  @Override
  public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor data) {
    MatrixCursor extras = new MatrixCursor(new String[]{
        KEY_ROWID,
        KEY_LABEL,
        KEY_CURRENCY
    });
    extras.addRow(new String[]{
        "0",
        getString(R.string.menu_create_account),
        Utils.getHomeCurrency().getCode()
    });
    mAccountsCursor = new MergeCursor(new Cursor[]{extras, data});
    mAccountsAdapter.swapCursor(mAccountsCursor);
    UiUtils.selectSpinnerItemByValue(mAccountSpinner, accountId);
  }

  @Override
  public void onLoaderReset(@NonNull Loader<Cursor> loader) {
    mAccountsCursor = null;
    mAccountsAdapter.swapCursor(null);
  }

  @Override
  protected void setupDialogView(View view) {
    super.setupDialogView(view);
    mAccountSpinner = view.findViewById(R.id.Account);
    Context wrappedCtx = view.getContext();
    mAccountsAdapter = new SimpleCursorAdapter(wrappedCtx, android.R.layout.simple_spinner_item, null,
        new String[]{KEY_LABEL}, new int[]{android.R.id.text1}, 0);
    mAccountsAdapter.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item);
    mAccountSpinner.setAdapter(mAccountsAdapter);
    mAccountSpinner.setOnItemSelectedListener(this);
    LoaderManager.getInstance(this).initLoader(0, null, this);

    mDateFormatSpinner = view.findViewById(R.id.DateFormat);
    DialogUtils.configureDateFormat(mDateFormatSpinner, wrappedCtx, prefHandler, PREF_KEY_IMPORT_DATE_FORMAT);

    mEncodingSpinner = view.findViewById(R.id.Encoding);
    DialogUtils.configureEncoding(mEncodingSpinner, wrappedCtx, prefHandler, PREF_KEY_IMPORT_ENCODING);

    mCurrencySpinner = view.findViewById(R.id.Currency);
    DialogUtils.configureCurrencySpinner(mCurrencySpinner, this);
    currencyViewModel.getCurrencies().observe(this, currencies -> {
      final CurrencyAdapter adapter = (CurrencyAdapter) mCurrencySpinner.getAdapter();
      adapter.addAll(currencies);
      mCurrencySpinner.setSelection(adapter.getPosition(currencyViewModel.getDefault()));
    });
    view.findViewById(R.id.AccountType).setVisibility(View.GONE);//QIF data should specify type
  }

  @Override
  public void onItemSelected(AdapterView<?> parent, View view, int position,
                             long id) {
    if (parent.getId() == R.id.Currency) {
      if (accountId == 0) {
        currency = ((Currency) parent.getSelectedItem()).getCode();
      }
      return;
    }
    if (mAccountsCursor != null) {
      accountId = id;
      mAccountsCursor.moveToPosition(position);

      String currency = (accountId == 0 && this.currency != null) ?
          this.currency :
          mAccountsCursor.getString(2);//2=KEY_CURRENCY
      mCurrencySpinner.setSelection(
          ((ArrayAdapter<Currency>) mCurrencySpinner.getAdapter())
              .getPosition(Currency.Companion.create(currency, requireActivity())));
      mCurrencySpinner.setEnabled(position == 0);
    }
  }

  @Override
  public void onNothingSelected(AdapterView<?> parent) {
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    outState.putLong(KEY_ACCOUNTID, accountId);
    outState.putSerializable(KEY_CURRENCY, currency);
  }

}

File: myExpenses/src/main/java/org/totschnig/myexpenses/dialog/QifImportDialogFragment.kt
package org.totschnig.myexpenses.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.CheckBox
import android.widget.Spinner
import android.widget.TableRow
import androidx.appcompat.app.AlertDialog
import androidx.core.view.isVisible
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import kotlinx.coroutines.launch
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.QifImport
import org.totschnig.myexpenses.adapter.CurrencyAdapter
import org.totschnig.myexpenses.adapter.IdAdapter
import org.totschnig.myexpenses.export.qif.QifDateFormat
import org.totschnig.myexpenses.model.ExportFormat
import org.totschnig.myexpenses.provider.DatabaseConstants
import org.totschnig.myexpenses.util.ui.checkNewAccountLimitation
import org.totschnig.myexpenses.viewmodel.CurrencyViewModel
import org.totschnig.myexpenses.viewmodel.ImportConfigurationViewModel
import org.totschnig.myexpenses.viewmodel.data.AccountMinimal
import org.totschnig.myexpenses.viewmodel.data.Currency
import org.totschnig.myexpenses.viewmodel.data.Currency.Companion.create

class QifImportDialogFragment : TextSourceDialogFragment(), AdapterView.OnItemSelectedListener {
    private lateinit var accountSpinner: Spinner
    private lateinit var dateFormatSpinner: Spinner
    private lateinit var currencySpinner: Spinner
    private lateinit var encodingSpinner: Spinner
    private lateinit var autoFillRow: TableRow
    private lateinit var autoFillCategories: CheckBox
    @Suppress("UNCHECKED_CAST")
    private val accountsAdapter: IdAdapter<AccountMinimal>
        get() = accountSpinner.adapter as IdAdapter<AccountMinimal>
    private val currencyAdapter: CurrencyAdapter
        get() = currencySpinner.adapter as CurrencyAdapter
    private var currency: String? = null
    private val currencyViewModel: CurrencyViewModel by viewModels()
    private val viewModel: ImportConfigurationViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        with((requireActivity().application as MyApplication).appComponent) {
            inject(currencyViewModel)
            inject(viewModel)
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        if (savedInstanceState != null) {
            currency = savedInstanceState.getString(DatabaseConstants.KEY_CURRENCY)
        }
        return super.onCreateDialog(savedInstanceState)
    }

    override val layoutId = R.layout.import_dialog

    override val layoutTitle: String
        get() = getString(
            R.string.pref_import_title,
            format.name
        )

    private val format: ExportFormat
        get() = ExportFormat.QIF

    override val typeName = format.name

    override val prefKey = "import_" + format.extension + "_file_uri"

    override fun onClick(dialog: DialogInterface, id: Int) {
        if (activity == null) {
            return
        }
        if (id == AlertDialog.BUTTON_POSITIVE) {
            val format = dateFormatSpinner.selectedItem as QifDateFormat
            val encoding = encodingSpinner.selectedItem as String
            maybePersistUri()
            prefHandler.putString(PREF_KEY_IMPORT_ENCODING, encoding)
            prefHandler.putString(PREF_KEY_IMPORT_DATE_FORMAT, format.name)
            val currency = currencySpinner.selectedItem as? Currency
            if (currency != null) {
                (activity as QifImport).onSourceSelected(
                    uri!!,
                    format,
                    accountSpinner.selectedItemId,
                    currency.code,
                    mImportTransactions.isChecked,
                    mImportCategories.isChecked,
                    mImportParties.isChecked,
                    encoding,
                    autoFillCategories.isChecked
                )
            } else {
                showSnackBar("Currency is null")
            }
        } else {
            super.onClick(dialog, id)
        }
    }

    override fun setupDialogView(view: View) {
        super.setupDialogView(view)
        autoFillRow = view.findViewById(R.id.AutoFillRow)
        autoFillCategories = view.findViewById(R.id.autofill_categories)
        accountSpinner = view.findViewById(R.id.Account)
        accountSpinner.adapter = IdAdapter<AccountMinimal>(requireContext())
        accountSpinner.onItemSelectedListener = this
        dateFormatSpinner = view.findViewById(R.id.DateFormat)
        dateFormatSpinner.configureDateFormat(requireContext(), prefHandler, PREF_KEY_IMPORT_DATE_FORMAT)
        encodingSpinner = view.findViewById(R.id.Encoding)
        DialogUtils.configureEncoding(
            encodingSpinner,
            requireContext(),
            prefHandler,
            PREF_KEY_IMPORT_ENCODING
        )
        currencySpinner = view.findViewById(R.id.Currency)
        currencySpinner.configureCurrencySpinner(this)
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                currencyViewModel.currencies.collect {
                    val adapter = currencySpinner.adapter as CurrencyAdapter
                    adapter.addAll(it)
                    currencySpinner.setSelection(adapter.getPosition(currencyViewModel.default))
                }
            }
        }
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.accounts.collect {
                    accountsAdapter.clear()
                    accountsAdapter.addAll(it)
                    accountSpinner.setSelection(accountsAdapter.getPosition(viewModel.accountId))
                }
            }
        }

        view.findViewById<View>(R.id.AccountType).isVisible = false
        mImportTransactions.setOnCheckedChangeListener { _, isChecked ->
            autoFillRow.isVisible = isChecked
        }
    }

    override fun onItemSelected(
        parent: AdapterView<*>, view: View, position: Int,
        id: Long
    ) {
        if (parent.id == R.id.Currency) {
            if (viewModel.accountId == 0L) {
                currency = (parent.selectedItem as Currency).code
            }
            return
        }
        accountSpinner.checkNewAccountLimitation(prefHandler, requireContext())
        val selected = accountsAdapter.getItem(position)!!
        viewModel.accountId = selected.id
        val currency =
            if (selected.id == 0L && currency != null) currency else selected.currency
        currencySpinner.setSelection(
            currencyAdapter.getPosition(create(currency!!, requireActivity()))
        )
        currencySpinner.isEnabled = position == 0
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {}
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(DatabaseConstants.KEY_CURRENCY, currency)
    }

    companion object {
        const val PREF_KEY_IMPORT_DATE_FORMAT = "import_qif_date_format"
        const val PREF_KEY_IMPORT_ENCODING = "import_qif_encoding"

        @JvmStatic
        fun newInstance() = QifImportDialogFragment()
    }
}